### PR TITLE
Improve mod_pubsub test coverage

### PR DESCRIPTION
**Pull request by fenek**
New PR for rebased+squashed branch

### CODE REVIEW

File: apps/ejabberd/src/mod_pubsub.erl [hunk 1]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
...
+                                             NewPubSubPayload = #xmlel{ name = <<"pubsub">>, attrs = [{<<"xmlns">>, ?NS_PUBSUB_OWNER}],
+                                                                        children = [AffiliationsPayload] },
+                                             {error, {?ERR_NOT_ACCEPTABLE, [NewPubSubPayload]}};
+                                         _ ->
+                                             lists:foreach(fun ({JID, Affiliation}) ->
+                                                                   node_call(Host, Type, set_affiliation, [Nidx, JID, Affiliation]),
+                                                                   case Affiliation of
+                                                                       owner ->
+                                                                           NewOwner = jid:to_bare(JID),
+                                                                           NewOwners = [NewOwner | Owners],
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2721 is too long:                                                                            NewOwners = [NewOwner | Owners],.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 2]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
...
                                                                                      set_node,
                                                                                      [N#pubsub_node{owners = NewOwners}]);
+                                                                       none ->
+                                                                           OldOwner = jid:to_bare(JID),
+                                                                           case lists:member(OldOwner, Owners) of
+                                                                               true ->
+                                                                                   NewOwners = Owners -- [OldOwner],
+                                                                                   tree_call(Host,
+                                                                                             set_node,
+                                                                                             [N#pubsub_node{owners = NewOwners}]);
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2732 is too long:                                                                                              [N#pubsub_node{owners = NewOwners}]);.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 3]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
...
+                                                         ({JID, _}, Acc) ->
+                                                              sets:del_element(jid:to_bare(JID), Acc)
+                                                      end, sets:from_list(Owners), Entities),
+                                     case sets:size(OwnersDryRun) of
+                                         0 ->
+                                             OwnersPayload = [ #xmlel{ name = <<"affiliation">>,
+                                                                       attrs = [{<<"jid">>, jid:to_binary(Unchanged)}, {<<"affiliation">>, <<"owner">>}] }
+                                                               || Unchanged <- Owners ],
+                                             AffiliationsPayload = #xmlel{ name = <<"affiliations">>, attrs = [{<<"node">>, NodeId}],
+                                                                           children = OwnersPayload },
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2711 is too long:                                                                            children = OwnersPayload },.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 4]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
...
+                                             NewPubSubPayload = #xmlel{ name = <<"pubsub">>, attrs = [{<<"xmlns">>, ?NS_PUBSUB_OWNER}],
+                                                                        children = [AffiliationsPayload] },
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2713 is too long:                                                                         children = [AffiliationsPayload] },.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 5]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
         error ->
             {error, ?ERR_BAD_REQUEST};
         _ ->
-            Action = fun (#pubsub_node{type = Type, id = Nidx, owners = O} = N) ->
+            Action = fun (#pubsub_node{type = Type, id = Nidx, owners = O, nodeid = {_, NodeId}} = N) ->
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2693 is too long:             Action = fun (#pubsub_node{type = Type, id = Nidx, owners = O, nodeid = {_, NodeId}} = N) ->.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 6]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
...
+                                                                   node_call(Host, Type, set_affiliation, [Nidx, JID, Affiliation]),
+                                                                   case Affiliation of
+                                                                       owner ->
+                                                                           NewOwner = jid:to_bare(JID),
+                                                                           NewOwners = [NewOwner | Owners],
                                                                            tree_call(Host,
                                                                                      set_node,
                                                                                      [N#pubsub_node{owners = NewOwners}]);
+                                                                       none ->
+                                                                           OldOwner = jid:to_bare(JID),
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2726 is too long:                                                                            OldOwner = jid:to_bare(JID),.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 7]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
...
+                                                                       attrs = [{<<"jid">>, jid:to_binary(Unchanged)}, {<<"affiliation">>, <<"owner">>}] }
+                                                               || Unchanged <- Owners ],
+                                             AffiliationsPayload = #xmlel{ name = <<"affiliations">>, attrs = [{<<"node">>, NodeId}],
+                                                                           children = OwnersPayload },
+                                             NewPubSubPayload = #xmlel{ name = <<"pubsub">>, attrs = [{<<"xmlns">>, ?NS_PUBSUB_OWNER}],
+                                                                        children = [AffiliationsPayload] },
+                                             {error, {?ERR_NOT_ACCEPTABLE, [NewPubSubPayload]}};
+                                         _ ->
+                                             lists:foreach(fun ({JID, Affiliation}) ->
+                                                                   node_call(Host, Type, set_affiliation, [Nidx, JID, Affiliation]),
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2717 is too long:                                                                    node_call(Host, Type, set_affiliation, [Nidx, JID, Affiliation]),.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 8]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
...
-                                                               none ->
-                                                                   OldOwner = jid:to_lower(jid:to_bare(JID)),
-                                                                   case lists:member(OldOwner, Owners) of
-                                                                       true ->
-                                                                           NewOwners = Owners -- [OldOwner],
+                                     % It is a very simple check, as XEP doesn't state any
+                                     % other invalid affiliation transitions
+                                     OwnersDryRun = lists:foldl(
+                                                      fun({JID, owner}, Acc) ->
+                                                              sets:add_element(jid:to_bare(JID), Acc);
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2701 is too long:                                                               sets:add_element(jid:to_bare(JID), Acc);.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 9]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
...
+                                                                   case Affiliation of
+                                                                       owner ->
+                                                                           NewOwner = jid:to_bare(JID),
+                                                                           NewOwners = [NewOwner | Owners],
                                                                            tree_call(Host,
                                                                                      set_node,
                                                                                      [N#pubsub_node{owners = NewOwners}]);
+                                                                       none ->
+                                                                           OldOwner = jid:to_bare(JID),
+                                                                           case lists:member(OldOwner, Owners) of
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2727 is too long:                                                                            case lists:member(OldOwner, Owners) of.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 10]
```diff
@@ -2876,6 +2924,7 @@ get_subscriptions(Host, Node, JID, Plugins) when is_list(Plugins) ->
                                                 _ ->
                                                     []
                                             end;
+                                         %% used by node_flat (therefore by dag, hometree and push as well)
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2927 is too long:                                          %% used by node_flat (therefore by dag, hometree and push as well).

File: apps/ejabberd/src/mod_pubsub.erl [hunk 11]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
...
+                                     OwnersDryRun = lists:foldl(
+                                                      fun({JID, owner}, Acc) ->
+                                                              sets:add_element(jid:to_bare(JID), Acc);
+                                                         ({JID, _}, Acc) ->
+                                                              sets:del_element(jid:to_bare(JID), Acc)
+                                                      end, sets:from_list(Owners), Entities),
+                                     case sets:size(OwnersDryRun) of
+                                         0 ->
+                                             OwnersPayload = [ #xmlel{ name = <<"affiliation">>,
+                                                                       attrs = [{<<"jid">>, jid:to_binary(Unchanged)}, {<<"affiliation">>, <<"owner">>}] }
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2708 is too long:                                                                        attrs = [{<<"jid">>, jid:to_binary(Unchanged)}, {<<"affiliation">>, <<"owner">>}] }.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 12]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
...
+                                                               || Unchanged <- Owners ],
+                                             AffiliationsPayload = #xmlel{ name = <<"affiliations">>, attrs = [{<<"node">>, NodeId}],
+                                                                           children = OwnersPayload },
+                                             NewPubSubPayload = #xmlel{ name = <<"pubsub">>, attrs = [{<<"xmlns">>, ?NS_PUBSUB_OWNER}],
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2712 is too long:                                              NewPubSubPayload = #xmlel{ name = <<"pubsub">>, attrs = [{<<"xmlns">>, ?NS_PUBSUB_OWNER}],.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 13]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
...
+                                                              sets:add_element(jid:to_bare(JID), Acc);
+                                                         ({JID, _}, Acc) ->
+                                                              sets:del_element(jid:to_bare(JID), Acc)
+                                                      end, sets:from_list(Owners), Entities),
+                                     case sets:size(OwnersDryRun) of
+                                         0 ->
+                                             OwnersPayload = [ #xmlel{ name = <<"affiliation">>,
+                                                                       attrs = [{<<"jid">>, jid:to_binary(Unchanged)}, {<<"affiliation">>, <<"owner">>}] }
+                                                               || Unchanged <- Owners ],
+                                             AffiliationsPayload = #xmlel{ name = <<"affiliations">>, attrs = [{<<"node">>, NodeId}],
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2710 is too long:                                              AffiliationsPayload = #xmlel{ name = <<"affiliations">>, attrs = [{<<"node">>, NodeId}],.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 14]
```diff
@@ -2856,8 +2902,10 @@ get_subscriptions(Host, Node, JID, Plugins) when is_list(Plugins) ->
                                                 _ ->
                                                     []
                                             end;
+                                         %% no idea how to trigger this one
                                          ({_, none, _}) ->
                                             [];
+                                         %% sometimes used by node_pep
                                          ({#pubsub_node{nodeid = {_, SubsNode}}, Sub, SubId, SubJID}) ->
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2909 is too long:                                          ({#pubsub_node{nodeid = {_, SubsNode}}, Sub, SubId, SubJID}) ->.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 15]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
...
+                                                                           children = OwnersPayload },
+                                             NewPubSubPayload = #xmlel{ name = <<"pubsub">>, attrs = [{<<"xmlns">>, ?NS_PUBSUB_OWNER}],
+                                                                        children = [AffiliationsPayload] },
+                                             {error, {?ERR_NOT_ACCEPTABLE, [NewPubSubPayload]}};
+                                         _ ->
+                                             lists:foreach(fun ({JID, Affiliation}) ->
+                                                                   node_call(Host, Type, set_affiliation, [Nidx, JID, Affiliation]),
+                                                                   case Affiliation of
+                                                                       owner ->
+                                                                           NewOwner = jid:to_bare(JID),
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2720 is too long:                                                                            NewOwner = jid:to_bare(JID),.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 16]
```diff
@@ -1526,7 +1556,7 @@ send_pending_auth_events(Request, Host, Node, Owner) ->
 send_authorization_request(#pubsub_node{nodeid = {Host, Node}, type = Type, id = Nidx, owners = O},
                            Subscriber) ->
     Lang = <<"en">>,
-    Stanza = #xmlel{name = <<"message">>, attrs = [],
+    Stanza = #xmlel{name = <<"message">>, attrs = [{<<"id">>, list_to_binary(randoms:get_string())}],
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 1559 is too long:     Stanza = #xmlel{name = <<"message">>, attrs = [{<<"id">>, list_to_binary(randoms:get_string())}],.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 17]
```diff
@@ -1005,6 +1034,8 @@ do_route(ServerHost, Access, Plugins, Host, From, To, Packet) ->
                                   of
                                       {result, IQRes} ->
                                           jlib:iq_to_xml(IQ#iq{type = result, sub_el = IQRes});
+                                      {error, {Error, NewPayload}} ->
+                                          jlib:make_error_reply(Packet#xmlel{ children = NewPayload }, Error);
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 1038 is too long:                                           jlib:make_error_reply(Packet#xmlel{ children = NewPayload }, Error);.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 18]
```diff
@@ -2876,6 +2924,7 @@ get_subscriptions(Host, Node, JID, Plugins) when is_list(Plugins) ->
                                                 _ ->
                                                     []
                                             end;
+                                         %% used by node_flat (therefore by dag, hometree and push as well)
                                          ({#pubsub_node{nodeid = {_, SubsNode}}, Sub, SubJID}) ->
                                             case Node of
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> The expression on line 2929 and column 45 is nested beyond the maximum level of 3.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 19]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
...
+                                         _ ->
+                                             lists:foreach(fun ({JID, Affiliation}) ->
+                                                                   node_call(Host, Type, set_affiliation, [Nidx, JID, Affiliation]),
+                                                                   case Affiliation of
+                                                                       owner ->
+                                                                           NewOwner = jid:to_bare(JID),
+                                                                           NewOwners = [NewOwner | Owners],
                                                                            tree_call(Host,
                                                                                      set_node,
                                                                                      [N#pubsub_node{owners = NewOwners}]);
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2724 is too long:                                                                                      [N#pubsub_node{owners = NewOwners}]);.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 20]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
         error ->
             {error, ?ERR_BAD_REQUEST};
         _ ->
-            Action = fun (#pubsub_node{type = Type, id = Nidx, owners = O} = N) ->
+            Action = fun (#pubsub_node{type = Type, id = Nidx, owners = O, nodeid = {_, NodeId}} = N) ->
                              Owners = node_owners_call(Host, Type, Nidx, O),
                              case lists:member(Owner, Owners) of
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> The expression on line 2695 and column 30 is nested beyond the maximum level of 3.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 21]
```diff
@@ -2856,8 +2902,10 @@ get_subscriptions(Host, Node, JID, Plugins) when is_list(Plugins) ->
                                                 _ ->
                                                     []
                                             end;
+                                         %% no idea how to trigger this one
                                          ({_, none, _}) ->
                                             [];
+                                         %% sometimes used by node_pep
                                          ({#pubsub_node{nodeid = {_, SubsNode}}, Sub, SubId, SubJID}) ->
                                             case Node of
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> The expression on line 2910 and column 45 is nested beyond the maximum level of 3.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 22]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
...
                                                                            tree_call(Host,
                                                                                      set_node,
                                                                                      [N#pubsub_node{owners = NewOwners}]);
+                                                                       none ->
+                                                                           OldOwner = jid:to_bare(JID),
+                                                                           case lists:member(OldOwner, Owners) of
+                                                                               true ->
+                                                                                   NewOwners = Owners -- [OldOwner],
+                                                                                   tree_call(Host,
+                                                                                             set_node,
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2731 is too long:                                                                                              set_node,.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 23]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
...
-                                                                   case lists:member(OldOwner, Owners) of
-                                                                       true ->
-                                                                           NewOwners = Owners -- [OldOwner],
+                                     % It is a very simple check, as XEP doesn't state any
+                                     % other invalid affiliation transitions
+                                     OwnersDryRun = lists:foldl(
+                                                      fun({JID, owner}, Acc) ->
+                                                              sets:add_element(jid:to_bare(JID), Acc);
+                                                         ({JID, _}, Acc) ->
+                                                              sets:del_element(jid:to_bare(JID), Acc)
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2703 is too long:                                                               sets:del_element(jid:to_bare(JID), Acc).

File: apps/ejabberd/src/mod_pubsub.erl [hunk 24]
```diff
@@ -2660,41 +2690,56 @@ set_affiliations(Host, Node, From, EntitiesEls) ->
...
+                                                                           NewOwner = jid:to_bare(JID),
+                                                                           NewOwners = [NewOwner | Owners],
                                                                            tree_call(Host,
                                                                                      set_node,
                                                                                      [N#pubsub_node{owners = NewOwners}]);
+                                                                       none ->
+                                                                           OldOwner = jid:to_bare(JID),
+                                                                           case lists:member(OldOwner, Owners) of
+                                                                               true ->
+                                                                                   NewOwners = Owners -- [OldOwner],
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 2729 is too long:                                                                                    NewOwners = Owners -- [OldOwner],.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 25]
```diff
@@ -3742,11 +3793,12 @@ set_configure(Host, Node, From, Els, Lang) ->
                                                                end,
                                                      case set_xoption(Host, XData, OldOpts) of
                                                          NewOpts when is_list(NewOpts) ->
+                                                             NewNode = N#pubsub_node{options = NewOpts},
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 3796 is too long:                                                              NewNode = N#pubsub_node{options = NewOpts},.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 26]
```diff
@@ -3758,7 +3810,7 @@ set_configure(Host, Node, From, Els, Lang) ->
                                      end
                              end,
                     case transaction(Host, Node, Action, transaction) of
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> The expression on line 3812 and column 21 is nested beyond the maximum level of 3.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 27]
```diff
@@ -3742,11 +3778,12 @@ set_configure(Host, Node, From, Els, Lang) ->
                                                                end,
                                                      case set_xoption(Host, XData, OldOpts) of
                                                          NewOpts when is_list(NewOpts) ->
+                                                             NewNode = N#pubsub_node{options = NewOpts},
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> Line 3781 is too long:                                                              NewNode = N#pubsub_node{options = NewOpts},.

File: apps/ejabberd/src/mod_pubsub.erl [hunk 28]
```diff
@@ -3758,7 +3795,7 @@ set_configure(Host, Node, From, Els, Lang) ->
                                      end
                              end,
                     case transaction(Host, Node, Action, transaction) of
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_pubsub.erl]: According to **Elvis**:
> The expression on line 3797 and column 21 is nested beyond the maximum level of 3.

**Comment by michalwski** [comment]
This is massive coverage improvement! Thanks a lot!

**Pull request merged**